src/statemachine.rs

```diff
@@ -275,16 +290,45 @@ impl StateMachine {
             callback.clone(),
             status,
             move |info, selector, status, alive| {
-                let mut dev = match Self::init_and_select(info, &selector, &status, false, alive) {
-                    None => {
-                        return;
-                    }
+                let mut dev = match Self::init_device(info, &selector) {
                     Some(dev) => dev,
+                    None => return,
+                };
+
+                let args = args.clone();
+
+                // Attempt to silently discover credentials. If we find useful credentials, then we
+                // can potentially skip a device selection tap. We may find acceptable credentials
+                // on several devices. In that case, the first device to ask for automatic
+                // selection will be selected.
```

Comment:
I don't fully understand the `auto_select` feature, yet. 
Why do want potentially racy device-selection? 
I figured, we filter out devices that provide no suitable credentials here, by telling the `DeviceSelector`, they are `NotAToken()`. Then the `DeviceSelector` can decide if only a single device was discovered, and thus the selection tap can be skipped, or if multiple are discovered and a selection tap is necessary.

Reply:
We can't send NotAToken, because we need to blink at least one of the connected tokens. So we would need a new `DeviceSelector` message like `NotAUsefulToken`.

As for why we want this, suppose a user has multiple tokens connected and consider these cases:
1) Only one token stores a valid credential, but they forget which.
2) Several tokens contain valid credentials, and the relying party will accept any of them.
3) Several tokens contain valid credentials, and the relying party will only accept some of them.
4) Several tokens contain valid credentials, and the user really wants to use Token 1.

I believe 1 and 2 are common, and `auto_select` improves user experience in both.
As for 3, the current patch is not perfect, but with a few changes `auto_select` will improve UX in this case too. We just need to check that the device is suitable for the request, e.g. by checking the user verification requirement, after `init` and before `wait_for_device_selector`.

I don't think there's anything we can do about case 4. IMO it's not worth sacrificing the UX improvements for cases 1, 2, and 3 for the whim of a user who could just unplug their other tokens.

Reply:
> We can't send NotAToken, because we need to blink at least one of the connected tokens.

Right, I hadn't considered this. 

My question was indeed mostly from the standpoint of 4.
I don't have a good overview right now, how this change works with resident keys / discoverable credentials. There, a user may have account A on token 1 and account B on token 2, and thus it's important which token is used.

I do like the idea of having `NotAUsefulToken`, where the `DeviceSelector` could basically keep track of them, instead of throwing them away immediately. If one or more useful tokens is found, it can cancel the not useful ones as usual. And if all devices report back as being not useful, it could just send a dummy blink or simply `Continue` to all of them. 
Maybe in a follow-up PR?

Reply:
I'm going to disable `auto_select` in `sign` for now. I want to get the AppID changes into Firefox 118, and I think you're right that this feature isn't ready to go.